File: src/data/courses/dsa/stacks/stack-basic-applications/reverse-string-stack/code/java/ReverseStringStack.java
// ============================================================
// Reverse String Using Stack
// ============================================================

import java.util.*;

// ============================================================
// Solution 1: Two-Pointer Swap
// Time: O(N) | Space: O(1)
// ============================================================
class Solution1 {
    public String reverseString(String s) {
        char[] arr = s.toCharArray();
        int l = 0, r = arr.length - 1;
        while (l < r) { char t = arr[l]; arr[l] = arr[r]; arr[r] = t; l++; r--; }
        return new String(arr);
    }
}

// ============================================================
// Solution 2: Stack — Push All, Pop All
// Time: O(N) | Space: O(N)
// ============================================================
class Solution2 {
    public String reverseString(String s) {
        Deque<Character> stack = new ArrayDeque<>();
        for (char c : s.toCharArray()) stack.push(c);

        StringBuilder sb = new StringBuilder();
        while (!stack.isEmpty()) sb.append(stack.pop());
        return sb.toString();
    }
}

File: src/data/courses/dsa/stacks/stack-basic-applications/reverse-string-stack/code/cpp/reverse_string_stack.cpp
#include <string>
#include <stack>
#include <algorithm>
using namespace std;
// ============================================================
// Solution 1: Two-pointer swap — O(N) Time, O(1) Space
// ============================================================
class Solution1 {
public:
    string reverseString(string s) {
        int l=0, r=s.size()-1;
        while(l<r) swap(s[l++],s[r--]);
        return s;
    }
};

// ============================================================
// Solution 2: Stack — push all, pop all — O(N) Time+Space
// ============================================================
class Solution2 {
public:
    string reverseString(string s) {
        stack<char> st;
        for(char c:s) st.push(c);
        string res;
        while(!st.empty()){res+=st.top();st.pop();}
        return res;
    }
};
